server/schedule/placement/rule_list.go

```diff
@@ -113,17 +134,27 @@ func buildRuleList(rules map[[2]string]*Rule) (ruleList, error) {
 			sr.deleteRule(p.rule)
 		}
 		if i == len(points)-1 || !bytes.Equal(p.key, points[i+1].key) {
+			var endKey []byte
+			if i != len(points)-1 {
+				endKey = points[i+1].key
+			}
+
 			// next key is different, push sr to rl.
 			rr := sr.rules
 			if len(rr) == 0 {
-				var endKey []byte
-				if i != len(points)-1 {
-					endKey = points[i+1].key
-				}
 				return ruleList{}, errs.ErrBuildRuleList.FastGenByArgs(fmt.Sprintf("no rule for range {%s, %s}",
 					strings.ToUpper(hex.EncodeToString(p.key)),
 					strings.ToUpper(hex.EncodeToString(endKey))))
 			}
+
+			err := checkRulesRaft(rr)
```

Comment:
I think should check `applyRules` instead

Reply:
how about now?

Reply:
em, I meant that you need to check rules after calling `prepareRulesForApply`

https://github.com/tikv/pd/blob/ddb5665c465a1f65d5541b317b2bfb94ab33d97d/server/schedule/placement/rule_list.go#L165)

The reason is that rules can be overridden. It is acceptable if there are 2 rules both specify a leader while one rule is overridden by the other.

Reply:
Got it, done.